src/main/java/bio/terra/cli/service/utils/SamService.java

```diff
@@ -366,7 +369,28 @@ public void addUserToResource(
           },
           SamService::isRetryable);
     } catch (ApiException | InterruptedException ex) {
-      throw new SystemException("Error adding user to SAM resource.", ex);
+      if (!(ex instanceof ApiException)
+          || (((ApiException) ex).getCode() != HttpStatusCodes.STATUS_CODE_BAD_REQUEST)) {
+        throw new SystemException("Error adding user to SAM resource.", ex);
+      }
+      logger.info("User not found in SAM. Trying to invite a new user.");
+
+      try {
+        // add to resource failed with Not Found error, now try to invite the user and add them to
+        // the resource again
+        logger.info("Inviting new user: {}", userEmail);
+        UserStatusDetails userStatusDetails = inviteUser(userEmail);
+        logger.info("Invited new user: {}", userStatusDetails);
+
+        HttpUtils.callWithRetries(
+            () -> {
+              resourcesApi.addUserToPolicy(resourceType, resourceId, resourcePolicyName, userEmail);
+              return null;
+            },
+            SamService::isRetryable);
+      } catch (ApiException | InterruptedException secondEx) {
+        throw new SystemException("Error adding user to SAM resource.", secondEx);
+      }
```

Comment:
nit: is it worth adding a helper function for this pattern? code the same from the `WorkspaceManagerService#grantIamRole`, but I guess that's going away.
Maybe a samService function that [checks if the user is registered](https://sam.dsde-dev.broadinstitute.org/#/Users/getUserIds) and uses the invite api if not, and we call that before other functions involving a different user? Does add a Sam round trip for the rarer case of the user not existing.
Alternatively, we could add a function with a signature like `HttpUtils.callWithRetries` for `SamService.retryWithInvitedUser` to do the bad request, not found, add the user, retry the first call.

Possibly all overkill at this point.

Reply:
- The `WorkspaceManagerService#grantIamRole` isn't going away, I don't think?
- I took a shot at defining this as a pattern (try something/handle one-time error/try again) and using it in the three places that do this in CLI (two with invite, one with register). I'm not sure whether it's a net positive though, so I'm going to put it up as a separate PR and see what you think.